sno/dataset2.py

```diff
@@ -122,6 +125,141 @@ def hexhash(self):
         return _hexhash(self.dumps())
 
 
+def _pk_index_ordering(column):
+    """Returns primary key columns first, in pk_index order, then other columns."""
+    if column.pk_index is not None:
+        return column.pk_index
+    else:
+        # Non primary-key columns after primary key columns.
+        return float('inf')
+
+
+class ColumnSchema:
+    """
+    The schema for a single column. A column has a unique ID that is constant for the columns lifetime
+    - even if the column is moved or renamed - and it has a name, a datatype, info about if this
+    column is one of the primary keys, and potentially extra info about the specific datatype.
+    """
+
+    def __init__(self, id, name, data_type, pk_index=None, extra_type_info=None):
+        self.id = id
+        self.name = name
+        self.data_type = data_type
+        self.pk_index = pk_index
+        self.extra_type_info = extra_type_info if extra_type_info is not None else {}
+
+    @classmethod
+    def from_json_dict(cls, json_dict):
+        return cls(
+            json_dict.pop("id"),
+            json_dict.pop("name"),
+            json_dict.pop("dataType"),
+            json_dict.pop("primaryKeyIndex"),
+            json_dict,
+        )
```

Comment:
```suggestion
        return cls(**json_dict)
```

Reply:
Done, except the JSON names don't line up with the python names - partly because of our mild convention for short python symbols, but also due to the strong python convention of underscore_case symbols